File: sequenceandstring/Anagrams.cpp
#include <vector>
#include <unordered_map>
#include <string>
#include <algorithm>

using namespace std;

//http://www.lintcode.com/en/problem/anagrams/
/*
Anagrams

Given an array of strings, return all groups of strings that are anagrams.

Notice

All inputs will be in lower-case

Example
Given ["lint", "intl", "inlt", "code"], return ["lint", "inlt", "intl"].

Given ["ab", "ba", "cd", "dc", "e"], return ["ab", "ba", "cd", "dc"].
*/
class SolutionAnagrams {
public:
	/**
	* @param strs: A list of strings
	* @return: A list of strings
	*/
	vector<string> anagrams(vector<string> &strs) {
		// write your code here
		unordered_map<string, string> map;
		auto end = map.end();
		string dummy = "A";
		vector<string> anag;
		for (auto str : strs)
		{
			string tmp = str;
			std::sort(str.begin(), str.end());
			if (end == map.find(str))
				map[str] = tmp;
			else
			{
				if (map[str] != dummy)
				{
					anag.emplace_back(map[str]);
					map[str] = dummy;
				}
				anag.emplace_back(tmp);
			}
		}
		return anag;
	}
};

File: java/Anagrams.java
public class SolutionAnagrams {
    /**
     * @param s: The first string
     * @param b: The second string
     * @return true or false
     */
    public boolean anagram(String s, String t) {
        // write your code here
        if(s == null || t == null)return false;
        int lens = s.length();
        int lent = t.length();
        if(lens != lent)return false;
        char[] map = new char[256];
        for(int i = 0; i < lens; ++i)
            map[(int)s.charAt(i)]++;
        for(int i = 0; i < lens; ++i) {
            if(map[(int)t.charAt(i)] == 0)return false;
            map[(int)t.charAt(i)]--;
        }
        return true;
    }
};
